pgxn/neon/file_cache_internal.h

```diff
@@ -0,0 +1,266 @@
+ * means doing PS reads, to make sure we don't drop changes for pages we're
+ * fetching through prewarm processes.
+ *
+ * The algorithm used is as follows:
+ *
+ * 1.  Select a chunk to prewarm.
+ *     This requires the chunk to be present in LFC, or an empty chunk to be
+ *     available in the LFC that we can then fill.
+ *     If there are no empty chunks left, and the chunk we selected isn't
+ *     present in the LFC, we skip this chunk as evicting data to prewarm other
+ *     data may be detrimental to the workload.
+ * 2.  Prepare the chunk: with exclusively locked LFC:
+ *     1.  Note down which blocks are already written to the chunk.
+ *         We don't have to read those blocks, as they're already the latest
+ *         version.
+ *         If we don't have any blocks left to fetch, we're done: restart at 1
+ *     2.  Unlink the selected chunk from the LRU.
+ *         Now it can't be removed through LRU eviction, so we know any new
+ *         blocks written to the chunk will be the latest version of said pages.
+ * 3.  Determine which blocks to retrieve (= selection from 1, with selection
+ *     from 2.1 removed)
+ * 4.  Prefetch the to-be-retrieved blocks of the chunk.
+ * 5.  Read all to-be-retrieved blocks of the chunk into local memory.
+ * 6.  With exclusively locked LFC:
+ *     1.  Mark chunk as LFC_PREWARM_STARTED
+ *     2.  Copy this backend's prewarmer worker number (if relevant) into new
+ *         `prewarm_worker_field` on chunk (this needs to be only a few bits,
+ *         e.g. 8, as we only will likely have very few workers concurrently
+ *         prewarming the LFC),
+ *     3.  Copy the LFC entry's refcount into per-prewarmworker state's
+ *         'wait_count', reduced by 1 (to account for the bgworker itself).
+ *     Now the chunk we're going to put data into is pinned (has been removed from the LRU, can't be reclaimed) and no new backends will start write-IO on pages that aren't present in the chunk due to the LFC_PREWARM_STARTED flag.
+ * 7.  Normal backends: Writes for not-yet-present pages that get to that chunk
+ *     start waiting on (new) condition variable `prewarm_done` in the
+ *     reported prewarm worker's worker state, until the lfc chunk doesn't have
+ *     `LFC_PREWARM_STARTED` set or the page is present in the LFC chunk. All
+ *     other writes and reads can continue as usual (those pages are not
+ *     touched by the prewarm worker).
+ * 8.  Normal backends: When the chunk is released:
+ *     1.  If LFC_PREWARM_STARTED is not set, we're done.
+ *     2.  If LFC_PREWARM_STARTED was already set when the chunk was acquired,
+ *         we're done.
+ *     3.  Reduce the current chunk's prewarm worker's wait count by 1.
+ *     4.  If the current chunk's prewarm worker's wait count is 0, signal
+ *         chunkrelease.
+ * 9.  Prewarmer:
+ *     1.  wait on new condition variable `chunkrelease` until the wait_count
```

Comment:
I am not sure if having `max_connections` condition variables (taking pin account not only prewarm case) is ok.
My be we should better maintain some turnstile of condition variables...